engine/chassis.ts

```diff
@@ -109,77 +434,325 @@ export class Program {
         return new ProgramInputs(this, info);
     }
 
-    getHandle (name: string, offset = 0, type = Type.UNKNOWN): IProgramHandle {
-        const handle = 0;
+    // related to uniform buffer
+    getHandle (name: string, offset = 0, type = Type.UNKNOWN, arrayIdx = 0): IProgramHandle {
+        let handle = this._propertyMap[name];
+        if (!handle) { return 0 as unknown as IProgramHandle; }
+        const memberType = getTypeFromHandle(handle);
+        if (type) {
+            handle = customizeType(handle, type);
+        } else if (offset) {
+            handle = customizeType(handle, getTypeFromHandle(handle) - offset);
+        }
+        if (arrayIdx > 0) {
+            handle += arrayIdx * (TypeInfos[memberType].size / 4);
+        }
+        handle += offset;
         return handle as unknown as IProgramHandle;
     }
 
     // update current state settings
     setPipelineState (info: IPipelineStateInfo) {
-        // render pass & input states are hard-coded for now
-        const hash = hashObject(info);
+        const stateHash = hashObject(info);
+
+        if (!this._pipelineStateMap.has(stateHash)) {
+            this._pipelineStateMap.set(stateHash, TestBase.device.createPipelineState(new PipelineStateInfo(
+                this._shader,
+                this._layout,
+                TestBase.defaultRenderPass,
+                new InputState(this._shader.attributes),
+                info.rasterizerState,
+                info.depthStencilState,
+                info.blendState,
+                info.primitive,
+            )));
+        }
+        this._currentPipelineState = this._pipelineStateMap.get(stateHash)!;
     }
 
-    draw (commandBuffer: CommandBuffer, bindings: ProgramBindings, inputs: ProgramInputs) {
+    draw (commandBuffer: CommandBuffer, bindings: ProgramBindings, inputs: ProgramInputs, instance?: number | undefined) {
+        TestBase.assert(!!this._currentPipelineState, `no pipelineState specified`);
+        commandBuffer.bindPipelineState(this._currentPipelineState);
+
+        if (instance === undefined) {
+            bindings.descriptorSets.map((descriptorSet, index) => commandBuffer.bindDescriptorSet(index, descriptorSet));
+        } else {
+            for (let i = 0; i < bindings.descriptorSets.length; i++) {
+                const offsets = bindings.dynamicOffsets[i].map((offset) => offset * instance);
```

Comment:
The `draw` function is the most frequently invoked routine, so be very careful doing the procedures required here: unnecessary lamdas and temporary variables in such places are common culprits for performance hits.